File: vortexdbg-android/src/main/java/com/vortexdbg/linux/android/dvm/ArmVarArg.java
package com.vortexdbg.linux.android.dvm;

import com.vortexdbg.Emulator;
import com.vortexdbg.pointer.VortexdbgPointer;

public abstract class ArmVarArg extends VarArg {

    static VarArg create(Emulator<?> emulator, BaseVM vm, DvmMethod method) {
        return emulator.is64Bit() ? new ArmVarArg64(emulator, vm, method) : new ArmVarArg32(emulator, vm, method);
    }

    protected final Emulator<?> emulator;

    protected ArmVarArg(Emulator<?> emulator, BaseVM vm, DvmMethod method) {
        super(vm, method);
        this.emulator = emulator;
    }

    private static final int REG_OFFSET = 3;

    protected final VortexdbgPointer getArg(int index) {
        return emulator.getContext().getPointerArg(REG_OFFSET + index);
    }

    protected final int getInt(int index) {
        VortexdbgPointer ptr = getArg(index);
        return ptr == null ? 0 : ptr.toIntPeer();
    }

}

File: vortexdbg-android/src/main/kotlin/com/vortexdbg/linux/android/dvm/ArmVarArg.kt
package com.vortexdbg.linux.android.dvm

import com.vortexdbg.Emulator
import com.vortexdbg.arm.context.RegisterContext
import com.vortexdbg.pointer.VortexdbgPointer

abstract class ArmVarArg protected constructor(protected val emulator: Emulator<*>, vm: BaseVM, method: DvmMethod) : VarArg(vm, method) {

    protected fun getArg(index: Int): VortexdbgPointer? {
        return emulator.getContext<RegisterContext>().getPointerArg(REG_OFFSET + index)
    }

    protected fun getInt(index: Int): Int {
        val ptr = getArg(index)
        return if (ptr == null) 0 else ptr.toIntPeer()
    }

    companion object {
        @JvmStatic
        fun create(emulator: Emulator<*>, vm: BaseVM, method: DvmMethod): VarArg {
            return if (emulator.is64Bit()) ArmVarArg64(emulator, vm, method) else ArmVarArg32(emulator, vm, method)
        }

        private const val REG_OFFSET = 3
    }
}
